release/serve_tests/workloads/locust_utils.py

```diff
@@ -0,0 +1,172 @@
+                        "response_time": response_time,
+                        "start_time": start_time,
+                    }
+                    self.errors.append(err_data)
+                    print(
+                        f"Request '{request_id}' failed with exception: {response.text}"
+                    )
+
+        class StagesShape(LoadTestShape):
+            def tick(self):
+                run_time = self.get_run_time()
+                for stage in stages:
+                    if run_time < stage["duration"]:
+                        tick_data = (stage["users"], stage["spawn_rate"])
+                        return tick_data
+                return None
+
+        self.user_class = EndpointUser
+        self.shape_class = StagesShape
+
+
+@ray.remote(num_cpus=1)
+class LocustWorker(LocustClient):
+    def __init__(
+        self,
+        host_url: str,
+        token: str,
+        master_address: str,
+        data: Dict[str, Any] = None,
+    ):
+        import locust
+        from locust.env import Environment
+        from locust.log import setup_logging
+
+        super().__init__(host_url=host_url, token=token, data=data)
+        setup_logging("INFO")
+        self.env = Environment(user_classes=[self.user_class], events=locust.events)
+        self.master_address = master_address
+
+    def run(self) -> List[Dict]:
+        runner = self.env.create_worker_runner(
+            master_host=self.master_address, master_port=5557
+        )
+        runner.greenlet.join()
+        return self.errors
+
+
+@ray.remote(num_cpus=1)
+class LocustMaster(LocustClient):
+    def __init__(
+        self,
+        host_url: str,
+        token: str,
+        expected_num_workers: int,
+        stages: List[Dict],
+    ):
+        from locust.env import Environment
+        from locust.log import setup_logging
+        import locust
+
+        super().__init__(host_url=host_url, token=token, stages=stages)
+        setup_logging("INFO")
+
+        self.master_env = Environment(
+            user_classes=[self.user_class],
+            shape_class=self.shape_class(),
+            events=locust.events,
+        )
+        self.expected_num_workers = expected_num_workers
+
+    def run(self):
+        import gevent
+        from locust.stats import (
+            get_stats_summary,
+            get_percentile_stats_summary,
+            get_error_report_summary,
+            stats_history,
+            stats_printer,
+        )
+
+        master_runner = self.master_env.create_master_runner("*", 5557)
+
+        while len(master_runner.clients.ready) < self.expected_num_workers:
+            print(
+                f"Waiting for workers to be ready, {len(master_runner.clients.ready)} "
+                f"of {self.expected_num_workers} ready."
+            )
+            time.sleep(1)
+
+        # Stats stuff
+        gevent.spawn(stats_printer(self.master_env.stats))
+        gevent.spawn(stats_history, master_runner)
```

Comment:
what do we need gevent for? would prefer not to introduce another dependency here. should we just use asyncio?

Reply:
gevent is a locust dependency. The locust CLI uses gevent to print out stats every 5-10 seconds, so I mimicked it here so that we can get stats printed out in the job.

Reply:
I see. seems fine. can you leave a better comment than "Stats stuff" :)